util/pkg/vfs/miniofs.go

```diff
@@ -0,0 +1,366 @@
+		}
+		minioEndpoint := os.Getenv("MINIO_ENDPOINT")
+		if minioEndpoint == "" {
+			return nil, fmt.Errorf("MINIO_ENDPOINT cannot be empty when KOPS_STATE_STORE is starting with minio")
+		}
+
+		s3Config := &aws.Config{
+			Credentials:      credentials.NewStaticCredentials(minioAccessKeyID, minioSecretAccessKey, ""),
+			Endpoint:         aws.String("http://" + minioEndpoint + ":9000"),
+			Region:           aws.String(region),
+			DisableSSL:       aws.Bool(true),
+			S3ForcePathStyle: aws.Bool(true),
+		}
+		newSession := session.New(s3Config)
+		s3Client = s3.New(newSession)
+	}
+
+	s.clients[region] = s3Client
+
+	return s3Client, nil
+}
+
+func (p *MinioPath) Path() string {
+	return "minio://" + p.bucket + "/" + p.key
+}
+
+func (p *MinioPath) Bucket() string {
+	return p.bucket
+}
+
+func (p *MinioPath) Key() string {
+	return p.key
+}
+
+func (p *MinioPath) String() string {
+	return p.Path()
+}
+
+func (p *MinioPath) Remove() error {
+	client, err := p.client()
+	if err != nil {
+		return err
+	}
+
+	request := &s3.DeleteObjectInput{}
+	request.Bucket = aws.String(p.bucket)
+	request.Key = aws.String(p.key)
+
+	_, err = client.DeleteObject(request)
+	if err != nil {
+		// TODO: Check for not-exists, return os.NotExist
+
+		return fmt.Errorf("error deleting %s: %v", p, err)
+	}
+
+	return nil
+}
+
+func (p *MinioPath) Join(relativePath ...string) Path {
+	args := []string{p.key}
+	args = append(args, relativePath...)
+	joined := path.Join(args...)
+	return &MinioPath{
+		minioContext: p.minioContext,
+		bucket:       p.bucket,
+		key:          joined,
+	}
+}
+
+func (p *MinioPath) WriteFile(data []byte) error {
+	client, err := p.client()
+	if err != nil {
+		return err
+	}
+
+	glog.V(4).Infof("Writing file %q", p)
+
+	// We always use server-side-encryption; it doesn't really cost us anything
+	sse := "AES256"
+
+	request := &s3.PutObjectInput{}
+	request.Body = bytes.NewReader(data)
+	request.Bucket = aws.String(p.bucket)
+	request.Key = aws.String(p.key)
+	request.ServerSideEncryption = aws.String(sse)
+
+	acl := os.Getenv("KOPS_STATE_Minio_ACL")
+	acl = strings.TrimSpace(acl)
+	if acl != "" {
+		glog.Infof("Using KOPS_STATE_Minio_ACL=%s", acl)
+		request.ACL = aws.String(acl)
+	}
+
+	// We don't need Content-MD5: https://github.com/aws/aws-sdk-go/issues/208
+
+	glog.V(8).Infof("Calling Minio PutObject Bucket=%q Key=%q SSE=%q ACL=%q BodyLen=%d", p.bucket, p.key, sse, acl, len(data))
+
+	_, err = client.PutObject(request)
+	if err != nil {
+		if acl != "" {
+			return fmt.Errorf("error writing %s (with ACL=%q): %v", p, acl, err)
+		} else {
+			return fmt.Errorf("error writing %s: %v", p, err)
+		}
+	}
+
+	return nil
+}
+
+// To prevent concurrent creates on the same file while maintaining atomicity of writes,
+// we take a process-wide lock during the operation.
+// Not a great approach, but fine for a single process (with low concurrency)
+// TODO: should we enable versioning?
```

Comment:
My understanding- we do enable versioning for s3. So it would be good to have it here as well. Need not be in this change.